consensus/src/block_storage/sync_manager.rs

```diff
@@ -248,16 +249,69 @@ impl BlockStore {
         quorum_certs.extend(
             blocks
                 .iter()
-                .take(num_blocks as usize - 1)
+                .take(blocks.len() - 1)
                 .map(|block| block.quorum_cert().clone()),
         );
+
+        // check if highest_commit_cert comes from a fork
+        // if so, we need to fetch it's block as well, to have a proof of commit.
+        if !blocks
+            .iter()
+            .any(|block| block.id() == highest_commit_cert.certified_block().id())
+        {
+            let mut additional_blocks = retriever
+                .retrieve_block_for_qc(
+                    highest_commit_cert,
+                    1,
+                    highest_commit_cert.commit_info().id(),
+                )
+                .await?;
+
+            assert_eq!(additional_blocks.len(), 1);
+            let block = additional_blocks.pop().expect("blocks are empty");
+            assert_eq!(
+                block.id(),
+                highest_commit_cert.certified_block().id(),
+                "Expecting in the retrieval response, for commit certificate fork, first block should be {}, but got {}",
+                highest_commit_cert.certified_block().id(),
+                block.id(),
+            );
+
+            blocks.push(block);
+            quorum_certs.push(highest_commit_cert.clone());
+        }
+
+        assert_eq!(blocks.len(), quorum_certs.len());
         for (i, block) in blocks.iter().enumerate() {
             assert_eq!(block.id(), quorum_certs[i].certified_block().id());
         }
 
         // If a node restarts in the middle of state synchronization, it is going to try to catch up
         // to the stored quorum certs as the new root.
         storage.save_tree(blocks.clone(), quorum_certs.clone())?;
+
+        // Check early that recovery will succeed, and return before corrupting our state in case it will not.
+        LedgerRecoveryData::new(highest_commit_cert.ledger_info().clone())
```

Comment:
this needs to be checked before it's saved to the db. also the & mut blocks.clone() are scary. let's do something more explicit like
let blocks_for_check = blocks.clone();